### PR TITLE
mgr/volumes: Better handle config file on osd full scenario

### PR DESCRIPTION
**Pull request by kotreshhr**
The 'metadata_mgr.flush()' used to truncate the config file
before flushing the new config data. This could lead to an
empty config file when there is no space to write new config
data. This patch handles this scenario by writing it to
temporary file and rename it to config file. This would
retain the config file without truncating it.

Also, there are bunch of places which wasn't handling
'MetadataMgrException' because of this. Fixed those.

Fixes: https://tracker.ceph.com/issues/55976
Signed-off-by: Kotresh HR <khiremat@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by kotreshhr** [comment]
jenkins test api

**Comment by kotreshhr** [comment]
jenkins test windows

**Comment by kotreshhr** [comment]
fs:volumes test looks good:
http://pulpito.front.sepia.ceph.com/khiremat-2022-06-16_07:22:57-fs:volumes-wip-khiremat-46599-rename_subvolume_config-0-distro-default-smithi/

File: src/pybind/mgr/volumes/fs/operations/versions/metadata_manager.py [hunk 1]
```diff
@@ -75,15 +75,18 @@ def flush(self):
...
             while True:
                 data = conf_data.read()
                 if not len(data):
                     break
                 wrote += self.fs.write(fd, data.encode('utf-8'), -1)
             self.fs.fsync(fd, 0)
-            log.info("wrote {0} bytes to config {1}".format(wrote, self.config_path))
+            log.info("wrote {0} bytes to config {1}".format(wrote, tmp_config_path))
+            self.fs.rename(tmp_config_path, self.config_path)
+            log.info("Renamed {0} to config {1}".format(tmp_config_path, self.config_path))
```
Comment (mchangir) [inline src/pybind/mgr/volumes/fs/operations/versions/metadata_manager.py:89]: please use `f""` format strings if possible
they are easier on the eyes and easier to understand

File: src/pybind/mgr/volumes/fs/operations/versions/metadata_manager.py [hunk 2]
```diff
@@ -75,15 +75,18 @@ def flush(self):
 
         fd = None
         try:
-            fd = self.fs.open(self.config_path, os.O_WRONLY | os.O_CREAT | os.O_TRUNC, self.mode)
+            tmp_config_path = self.config_path + b'.tmp'
+            fd = self.fs.open(tmp_config_path, os.O_WRONLY | os.O_CREAT | os.O_TRUNC, self.mode)
```
Comment (mchangir) [inline src/pybind/mgr/volumes/fs/operations/versions/metadata_manager.py:79]: Could attempting to create a temporary file fail as well ?
Is it possible to open the desired path (`self.config_path`) and then use `self.fs.fallocate(fd, ...)` to ensure availability of space an then write to it directly ?
You might have to explicitly truncate the `self.config_path` to the new size anyway.

File: src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py [hunk 3]
```diff
@@ -742,9 +746,14 @@ def remove_snapshot(self, snapname):
         if self.has_pending_clones(snapname):
             raise VolumeException(-errno.EAGAIN, "snapshot '{0}' has pending clones".format(snapname))
         snappath = self.snapshot_path(snapname)
+        try:
+            self.metadata_mgr.remove_section(self.get_snap_section_name(snapname))
```
Comment (kotreshhr) [inline src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:750]: @vshankar @nmshelke @ajarr 

The order is reversed here. Previously, snapshot was removed first and then the metadata, now the other way. This is the right thing to do as we  end up keeping the metadata of the deleted snapshot if metadata deletion fails. 

But there is a side effect of doing this. The metadata deletion usually fails with ENOSPACE (OSD full scenario). So when there is no space, the snapshot deletion fails with new code. But with old code, the snapshot deletion is successful but the metadata deletion fails. Though the older behavior is not correct, it helps in recovering space during OSD full scenario through subvolume APIs.

We need to think and come up with some sort of solution to make snapshot deletion goes through when there is no space.  One solution is to allow the older behavior with `force` option. But we have to think of a solution to clean up the stale snapshot metadata. 

Thought ?

Comment (vshankar) [reply src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:750]: I recall that we discussed this, did we decide on keeping the old behavior (delete snapshot first) and if the metadata update fails, we still report success and later when loading the subvolume, cleanup the metadata for which snaps are not present but its metadata exist.

Comment (kotreshhr) [reply src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:750]: Yes, we discussed that in one to one meeting. We had discussed to keep the old behavior with `force` option. It's not exactly deleting the snapshot first but ignoring the metadata update failure and reporting success.

>  later when loading the subvolume, cleanup the metadata for which snaps are not present but its metadata exist?
Is this scalable ? If there are thousands of snapshots per subvolume, we end up stating thousand+ times for every such subvolume load.

Comment (vshankar) [reply src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:750]: > Yes, we discussed that in one to one meeting. We had discussed to keep the old behavior with `force` option. It's not exactly deleting the snapshot first but ignoring the metadata update failure and reporting success.
> 
> > later when loading the subvolume, cleanup the metadata for which snaps are not present but its metadata exist?
> 
> Is this scalable ? If there are thousands of snapshots per subvolume, we end up stating thousand+ times for every such subvolume load.

I recall now -- we thought about maintaining an index but that wasn't feasible either (w/ ENOSPC).

Comment (vshankar) [reply src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:750]: > @vshankar @nmshelke @ajarr
> 
> The order is reversed here. Previously, snapshot was removed first and then the metadata, now the other way. This is the right thing to do as we end up keeping the metadata of the deleted snapshot if metadata deletion fails.
> 
> But there is a side effect of doing this. The metadata deletion usually fails with ENOSPACE (OSD full scenario). So when there is no space, the snapshot deletion fails with new code. But with old code, the snapshot deletion is successful but the metadata deletion fails. Though the older behavior is not correct, it helps in recovering space during OSD full scenario through subvolume APIs.

This is more and more pushing me to think about ditching .meta and putting the metadata in a actual db (sqlite).

> 
> We need to think and come up with some sort of solution to make snapshot deletion goes through when there is no space. One solution is to allow the older behavior with `force` option. But we have to think of a solution to clean up the stale snapshot metadata.
> 
> Thought ?

Comment (kotreshhr) [reply src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:750]: > > Yes, we discussed that in one to one meeting. We had discussed to keep the old behavior with `force` option. It's not exactly deleting the snapshot first but ignoring the metadata update failure and reporting success.
> > > later when loading the subvolume, cleanup the metadata for which snaps are not present but its metadata exist?
> > 
> > 
> > Is this scalable ? If there are thousands of snapshots per subvolume, we end up stating thousand+ times for every such subvolume load.
> 
> I don't think we need to do a stat; can't we just list the .snap directory and then compare to the snap names? That should scale just fine.


@vshankar @gregsfortytwo 
Listing the .snap directory is time consuming on scale. I increased the `mds_max_snaps_per_dir` to maximum allowed i.e., 4096 . The listdir of 4096 snapshots on vstart cluster took 20+ seconds. If we choose to clean up stale snap metadata in subvolume discover, every subvolume API would consume so much time which is unacceptable.

Well, it works well with the default `mds_max_snaps_per_dir` i.e., 100. Listing 100 snaps is fast enough.

I am not really sure on what's the avg number of snapshots per subvolume in general but the solution to fix in subvolume discover doesn't scale well with allowed max snapshot limit.

How about leaving the stale snapshot metadata to stay around and clean it up on snapshot creation with the same name ? This way, the stale snapshot metadata stays and goes away in following two cases.

1.  When the subvolume is deleted
2.  When the snapshot is created with the same name in the corresponding subvolume.

File: src/pybind/mgr/volumes/fs/operations/versions/metadata_manager.py (hunk 2 above)
Comment (kotreshhr) [reply src/pybind/mgr/volumes/fs/operations/versions/metadata_manager.py:79]: I meant `fallocate` was tried here [1] but configparser wasn't friendly with it. 

[1] https://github.com/ceph/ceph/pull/44347#discussion_r827822449

**Comment by kotreshhr** [comment]
jenkins test windows

File: qa/workunits/fs/full/subvolume_clone.sh [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+# This testcase tests the  'ceph fs subvolume snapshot clone' when the osd is full.
+# The clone fails with 'MetadataMgrException: -28 (error in write)' and
+# truncates the config file of corresponding subvolume while updating the config file.
+# Hence the subsequent subvolume commands on the clone fails with
+# 'MetadataMgrException: -2 (section 'GLOBAL' does not exist)' traceback.
+
+# The osd is of the size 1GB. The full-ratios are set so that osd is treated full
+# at around 600MB. The subvolume is created and 100MB is written.
+# The subvolume is snapshotted and clone ten times. Since the clone delay is set to 15 seconds,
```
Comment (ajarr) [inline qa/workunits/fs/full/subvolume_clone.sh:12]: s/clone/cloned/

File: src/pybind/mgr/volumes/fs/volume.py [hunk 5]
```diff
@@ -496,7 +496,7 @@ def remove_subvolume_snapshot(self, **kwargs):
             with open_volume(self, volname) as fs_handle:
                 with open_group(fs_handle, self.volspec, groupname) as group:
                     with open_subvol(self.mgr, fs_handle, self.volspec, group, subvolname, SubvolumeOpType.SNAP_REMOVE) as subvolume:
-                        subvolume.remove_snapshot(snapname)
+                        subvolume.remove_snapshot(snapname, force)
```
Comment (ajarr) [inline src/pybind/mgr/volumes/fs/volume.py:499]: Is the "force" snapshot removal option being tested? Do we need a new test for this?

**Review by ajarr — CHANGES_REQUESTED** [review]
The commit message "https://github.com/ceph/ceph/pull/46599/commits/a28248b20d3ed8857b32382fe5c578254f4d11f9" has a typo, s/metadaa/metadata/ .  Also can you explain why we want to remove stale snapshot user metadata after discover() in the commit message body?

File: src/pybind/mgr/volumes/fs/operations/versions/subvolume_base.py [hunk 6]
```diff
@@ -407,9 +407,14 @@ def info(self):
...
     def set_user_metadata(self, keyname, value):
-        self.metadata_mgr.add_section(MetadataManager.USER_METADATA_SECTION)
-        self.metadata_mgr.update_section(MetadataManager.USER_METADATA_SECTION, keyname, str(value))
-        self.metadata_mgr.flush()
+        try:
+            self.metadata_mgr.add_section(MetadataManager.USER_METADATA_SECTION)
+            self.metadata_mgr.update_section(MetadataManager.USER_METADATA_SECTION, keyname, str(value))
+            self.metadata_mgr.flush()
+        except MetadataMgrException as me:
+            log.error(f"Failed to set user metadata key={keyname} value={value} on subvolume={self.subvol_name} "
```
[the remark this answers is not in the document]
Comment (kotreshhr) [reply src/pybind/mgr/volumes/fs/operations/versions/subvolume_base.py:419]: The exception is expected only in the cases where the metadata is being flushed when OSD is full. We don't flush metadata in `get_user_metadata()`, so we should be good.

File: src/pybind/mgr/volumes/fs/operations/versions/subvolume_base.py [hunk 7]
```diff
@@ -407,9 +407,14 @@ def info(self):
...
+                      f"group={self.group_name} reason={me.args[1]}, errno:{-me.args[0]}, {os.strerror(-me.args[0])}")
```
[the remark this answers is not in the document]
Comment (kotreshhr) [reply src/pybind/mgr/volumes/fs/operations/versions/subvolume_base.py:420]: We don't do that for any of the logs and the volume name is not stored in subvolume object as well. It's only used to fetch the fs_handle. I thought having groupname and subvolume name should be good enough to identify. Do you think it's necessary to store the volume just for this logging purpose ?

File: qa/workunits/fs/full/subvolume_clone.sh (hunk 4 above)
Comment (kotreshhr) [reply qa/workunits/fs/full/subvolume_clone.sh:12]: Done

File: src/pybind/mgr/volumes/fs/volume.py (hunk 5 above)
Comment (kotreshhr) [reply src/pybind/mgr/volumes/fs/volume.py:499]: Done

File: src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py [hunk 8]
```diff
@@ -792,6 +792,21 @@ def list_snapshots(self):
...
+        """ Clean up stale snapshot metadata """
+        try:
+            if self.metadata_mgr.has_snap_metadata_section():
+                snap_list = self.list_snapshots()
+                snaps_with_metadata_list = self.metadata_mgr.list_snaps_with_metadata()
+                for snap_with_metadata in snaps_with_metadata_list:
+                    if snap_with_metadata.encode('utf-8') not in snap_list:
+                        self.metadata_mgr.remove_section(self.get_snap_section_name(snap_with_metadata))
+                        self.metadata_mgr.flush()
+        except MetadataMgrException as me:
```
[the remark this answers is not in the document]
Comment (kotreshhr) [reply src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:805]: Done.

File: src/pybind/mgr/volumes/fs/operations/versions/subvolume_base.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ajarr) [reply src/pybind/mgr/volumes/fs/operations/versions/subvolume_base.py:419]: It wasn't obvious to me that you wanted to log errors only only when metadata_mgr.flush() . I assumed you wanted to log errors during different user metadata operations.

File: src/pybind/mgr/volumes/fs/operations/versions/metadata_manager.py [hunk 9]
```diff
@@ -75,15 +75,18 @@ def flush(self):
 
         fd = None
         try:
-            fd = self.fs.open(self.config_path, os.O_WRONLY | os.O_CREAT | os.O_TRUNC, self.mode)
+            tmp_config_path = self.config_path + b'.tmp'
```
[the remark this answers is not in the document]
Comment (ajarr) [reply src/pybind/mgr/volumes/fs/operations/versions/metadata_manager.py:79]: I guess flush() executed by only one thread at a time? I was concerned about the .tmp file being truncated by another thread just before being renamed.

File: src/pybind/mgr/volumes/fs/operations/versions/__init__.py [hunk 10]
```diff
@@ -95,7 +95,10 @@ def get_subvolume_object(self, mgr, fs, vol_spec, group, subvolname, upgrade=Tru
             subvolume.discover()
             self.upgrade_to_v2_subvolume(subvolume)
             version = int(subvolume.metadata_mgr.get_global_option('version'))
-            return self._get_subvolume_version(version)(mgr, fs, vol_spec, group, subvolname, legacy=subvolume.legacy_mode)
+            subvolume_version_object = self._get_subvolume_version(version)(mgr, fs, vol_spec, group, subvolname, legacy=subvolume.legacy_mode)
+            subvolume_version_object.metadata_mgr.refresh()
+            subvolume_version_object.clean_stale_snapshot_metadata()
```
Comment (ajarr) [inline src/pybind/mgr/volumes/fs/operations/versions/__init__.py:100]: can it be tested that stale snapshot metadata gets cleaned up?

**Review by ajarr — APPROVED** [review]

File: src/pybind/mgr/volumes/fs/operations/versions/subvolume_base.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (kotreshhr) [reply src/pybind/mgr/volumes/fs/operations/versions/subvolume_base.py:420]: Yes, creating a tracker ticket to address that sounds good to me.

File: src/pybind/mgr/volumes/fs/operations/versions/__init__.py (hunk 10 above)
Comment (kotreshhr) [reply src/pybind/mgr/volumes/fs/operations/versions/__init__.py:100]: Sure I will add the test case for this.

Comment (kotreshhr) [reply src/pybind/mgr/volumes/fs/operations/versions/__init__.py:100]: Done

**Comment by kotreshhr** [comment]
jenkins test windows

**Comment by kotreshhr** [comment]
jenkins test make check

**Comment by kotreshhr** [comment]
@vshankar PTAL

**Comment by vshankar** [comment]
LGTM

**Comment by vshankar** [comment]
(including in a separate branch since it conflicts with other PRs in wip-vshankar-testing branch)

**Comment by vshankar** [comment]
> This pull request can no longer be automatically merged: a rebase is needed and changes have to be manually resolved

Just FYI -- I have these changes under tests here: https://pulpito.ceph.com/vshankar-2022-08-12_09:34:24-fs-wip-vshankar-testing1-20220812-072441-testing-default-smithi/

**Comment by vshankar** [comment]
@kotreshhr please rebase. I have test running here https://pulpito.ceph.com/vshankar-2022-08-16_04:28:55-fs:volumes-wip-vshankar-testing1-20220812-072441-testing-default-smithi/ which has failures. Could you please take a look at those too?

**Comment by kotreshhr** [comment]
> @kotreshhr please rebase. I have test running here https://pulpito.ceph.com/vshankar-2022-08-16_04:28:55-fs:volumes-wip-vshankar-testing1-20220812-072441-testing-default-smithi/ which has failures. Could you please take a look at those too?

@vshankar Sure. The test failure is because of Rishab's changes w.r. t. sudo. I will fix that too.

**Comment by kotreshhr** [comment]
@vshankar This PR also needs a run of fs:full suite. Did that pass ?

**Comment by kotreshhr** [comment]
jenkins test windows

**Comment by kotreshhr** [comment]
jenkins test make check arm64

**Comment by vshankar** [comment]
Tests are fine -- waiting for jenkins issue to get fixed.

**Comment by kotreshhr** [comment]
jenkins test make check arm64

**Comment by kotreshhr** [comment]
jenkins test windows

**Comment by kotreshhr** [comment]
jenkins test windows

**Review by vshankar — APPROVED** [review]
https://tracker.ceph.com/projects/cephfs/wiki/Main#2022-Aug-22

**Pull request merged**